V10

tbillington opened the pull request.

tbillington committed:
Removing v9 api

tbillington committed:
Adding V10

tbillington committed:
Fixed Patch only taking 1 meta key value per doc

jonathaningram commented on src/sajari.php:
```diff
@@ -0,0 +1,442 @@
+require dirname(__FILE__) . '/document.php';
+require dirname(__FILE__) . '/search.php';
+
+interface Opt
```
@tbillington unless I'm missing changes in PHP these days, per PSR0 autoloading, every class and interface should be in its own file under a directory that matches the namespace. So WithEndpoint should be found at `src/Sajari/WithEndpoint.php`. Note: a newer PSR autoloading spec (PSR4?) lets you omit the `Sajari` directory I believe so the files can be found at `src/WithEndpoint.php`. Best check those specs though as it's been a while for me.

cc @dhowden

tbillington replied:
Thanks @jonathaningram that's the kind of domain knowledge I need 👍  I'll check it out.

jonathaningram replied:
@tbillington np. Also this is the best PHP has for "go fmt", if you haven't done so already try and set it up https://github.com/FriendsOfPHP/PHP-CS-Fixer e.g. maybe you can set it up as a pre-commit hook to auto-fmt or reject the commit if it shows warnings. Note: I tried quickly to get it working with Atom, but I don't think it's working for some reason.

jonathaningram commented on src/composer.json:
```diff
@@ -0,0 +1,14 @@
+{
```
@tbillington `composer.json` should be in the root dir. Also refer to `.gitignore` changes above which should move back to root dir (unless you did it for a particular reason and I'm missing something?).

tbillington replied:
I'll move it back, it was due to my lack of knowledge of correct php file structure.

jonathaningram commented on src/composer.json:
```diff
@@ -0,0 +1,14 @@
+  "repositories": [
+    {
+      "type": "vcs",
+      "url": "https://github.com/stanley-cheung/Protobuf-PHP"
```
@tbillington is this Github repo not on Packagist.org? Why does it need a custom `repositories` definition?

jonathaningram commented on src/composer.json:
```diff
@@ -0,0 +1,14 @@
+  "require": {
+    "grpc/grpc": "dev-release-0_14"
+  }
+}
```
@tbillington newline is missing

jonathaningram commented on src/document.php:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    /**
+     * @return \sajari\engine\store\doc\Keys\Key
```
@tbillington you'd need to check, but PHP namespaces with upper and lower case may have issues. `Sajari` and `sajari` I mean. Is it possible to have the auto-gen'd proto stuff use namespace `Sajari`? Further, can it make the namespaces first letter uppercase?

It probably wouldn't be a problem if the proto stuff was never exposed to the public API, but it looks like in this method for example, that it is exposed.

tbillington replied:
The proto should never be exposed. The problem is I don't see a nice way to have the proto stuff available on all the objects (to make it easy to construct requests) but not expose it to the end user.

jonathaningram commented on src/composer.json:
```diff
@@ -0,0 +1,14 @@
+    }
+  ],
+  "require": {
+    "grpc/grpc": "dev-release-0_14"
```
@tbillington you can put `php` in the require too so that the min PHP version is specified. Since you use array notation `[]` which is new in PHP 5.5 (I think) you might want to specify the min version. Or, at least make sure it's specified in the README.md

jonathaningram commented on examples/add.php:
```diff
@@ -0,0 +1,27 @@
+$c = new Client('myproject', 'mycollection', $opts);
+
+$d = new Document([
+    new Meta("_body", "main body of document"),
```
@tbillington single quotes for PHP (php-cs-fixer should fix this anyway).

jonathaningram commented on examples/add.php:
```diff
@@ -0,0 +1,27 @@
+]);
+
+try {
+    $docKey = $c->Add($d);
```
@tbillington lowercase `add` for method name.

jonathaningram commented on examples/delete.php:
```diff
@@ -0,0 +1,24 @@
+<?php
+
+require  dirname(__FILE__) . '/../src/sajari.php';
```
@tbillington for all the examples, once you put the classes/interfaces into their own files, you won't be able to just `require` one file (because there will be too many). In the v9 SDK it was annoying using composer to require Sajari for my examples, so you're thinking is right to include the _real source_ so maybe you can look at how some other repos require "themselves" in their examples. Let me know if my comments are not clear here.

jonathaningram commented on examples/delete.php:
```diff
@@ -0,0 +1,24 @@
+$key = new Key("_id", "7f7a8658-a368-d6f9-5e19-791debb6bddb");
+
+try {
+    $c->Delete($key);
```
@tbillington lowercase for `Delete` (and also for other examples)

I like the public API in these examples so good work. I'm just thinking though, what if you miss out the line 

```
$opts = [new WithEndpoint('server_address')];
```

? What happens if the client has no endpoint specified and would you ever want to do this?

jonathaningram replied:
@tbillington actually scrap my `WithEndpoint` question - it just uses the Sajari endpoint, duh.

jonathaningram replied:
@tbillington although, maybe that indicates in the examples it might be worth mentioning that they should only specify that if they are using a custom endpoint. Otherwise people will be asking what value they should put for the endpoint.

jonathaningram commented on examples/get.php:
```diff
@@ -0,0 +1,22 @@
+
+try {
+    $doc = $c->Get($key);
+} catch (Exception $e) {
```
@tbillington people will want to be able to handle engine exceptions differently to the global `Exception`, so if possible, make sure that there's a nice way for this. Some options:
1. Make the code throw `EngineException`s so they can be caught separately.
2. Add some static helper methods like `Sajari::isEngineError($e)`, `Sajari::isKeyConflictError($e)` per the "don’t assert errors for type, assert for behaviour" recommendation in http://dave.cheney.net/2014/12/24/inspecting-errors

jonathaningram commented on examples/search.php:
```diff
@@ -0,0 +1,101 @@
+$r->setFilter(
+    Sajari\All([
+        Sajari\EqualTo("type", "webpage"),
+        Sajari\EqualTo("category", "education"),
```
@tbillington filters look great. Maybe can put in a `pageviews > x` and a `not` just to show another operator or two?

jonathaningram commented on examples/search.php:
```diff
@@ -0,0 +1,101 @@
+]);
+
+try {
+    $res = $c->Search($r);
```
@tbillington search looks epic, nice.

tbillington committed:
Restructured to fit PSR-0

jonathaningram commented on src/Sajari/Client/Client.php:
```diff
@@ -0,0 +1,395 @@
+<?php
```
@tbillington I think this namespace should just be `Sajari` i.e. file exists at `src/Sajari/Client.php` - no need to stutter the `Client`, yeah?

jonathaningram commented on src/Sajari/Search/FilterMetaBoost.php:
```diff
@@ -0,0 +1,58 @@
+
+namespace Sajari\Search;
+
+abstract class MetaBoost
```
@tbillington separate file.

jonathaningram commented on src/Sajari/proto/doc.php:
```diff
@@ -0,0 +1,909 @@
+
+namespace sajari\engine\store\doc {
+
+  class XEmpty extends \DrSlump\Protobuf\Message {
```
@tbillington re your autoloading problem: have you checked if the protobuf generator can be configured to generate the files for PSR-0?

If it can't, then another option that might work but you need to verify: I think you can specify in composer.json that when it finds a namespace `sajari\engine\store\doc` (for example), to look in directory `src/Sajari/proto/` (or maybe look at the doc.php file in there), and then it should find the class.

tbillington committed:
Cleaning namespace usage

tbillington committed:
Updated proto, switched to ssl, fixed namespacing

Some modifications were made to the generated proto. The Deserialize
function was added to the response messages because they weren’t being
found. I assume this is because the implementation of Message has
changed, and one of the libraries is not in sync.

tbillington committed:
Fixed file name case

tbillington committed:
Added README.md

tbillington committed:
Update grpc version

tbillington committed:
Update README.md to include sample composer.json

tbillington committed:
Add more specific dev version in composer example

tbillington committed:
Fixing deserialize on XEmpty to satisfy php strict

tbillington committed:
Fixing deserialize for php strict

jdeniz committed:
Update Client.php, removed var_dumps, corrected references to generic Exception class

tbillington committed:
Merge branch 'jdeniz-v10' into v10

tbillington committed:
Throw useful message with Exception

tbillington committed:
Fix patch not using proto Key

Patch was using the wrapper type instead of the proto type, it was missing a call to the Proto method on the Key

tbillington committed:
Adding some basic structure to README (Incomplete)

tbillington committed:
Fixing search

tbillington committed:
Fixed Get

tbillington committed:
Fix Add and Delete

tbillington committed:
Fix patch

tbillington committed:
Fix whitespace issues

tbillington committed:
Fix get expecting wrong key type

tbillington committed:
Merge branch 'fixes' into v10

tbillington committed:
Add posneg and click tracking, update readme

tbillington committed:
Add more descriptions and examples to readme

tbillington committed:
Add constants to GeoFieldBoost class

tbillington committed:
Make Distance and Element field boosts extend FieldBoost

tbillington committed:
Fix incorrect variable name in ElementFieldBoost

tbillington committed:
Change geo field boost to geo filter

tbillington committed:
Change addfieldboost to additivefieldboost

tbillington committed:
Make setter methods on request taking arrays enforce array type

tbillington committed:
Fixed search, document operations are not fixed yet

tbillington committed:
Fix Add

tbillington committed:
Change to MIT license

tbillington committed:
Move to new grpc endpoint

tbillington committed:
Merge branch 'adding-tracking' into v10

tbillington commented:
This pull request doesn't make sense now that master is v10.
